### ISSUE TITLE
Questions on the mcbn

### ISSUE THREAD

**Issue by lan-qing** [iclr-mcbn/mcbn#1]
Dear authors:

Thanks for your nice work and codes. I am trying to understand the work and codes. I meet several problems and hope you can help me.

The MCBN is implemented in Line 147-193, [code](https://github.com/icml-mcbn/mcbn/tree/master/code)/[cifar10](https://github.com/icml-mcbn/mcbn/tree/master/code/cifar10)/resnet_model.py:

```
def _batch_norm(self, name, x):
    """Batch normalization."""
    with tf.variable_scope(name):
      params_shape = [x.get_shape()[-1]]

      beta = tf.get_variable(
          'beta', params_shape, tf.float32,
          initializer=tf.constant_initializer(0.0, tf.float32))
      gamma = tf.get_variable(
          'gamma', params_shape, tf.float32,
          initializer=tf.constant_initializer(1.0, tf.float32))

      if self.mode == 'train':
        mean, variance = tf.nn.moments(x, [0, 1, 2], name='moments')

        moving_mean = tf.get_variable(
            'moving_mean', params_shape, tf.float32,
            initializer=tf.constant_initializer(0.0, tf.float32),
            trainable=False)
        moving_variance = tf.get_variable(
            'moving_variance', params_shape, tf.float32,
            initializer=tf.constant_initializer(1.0, tf.float32),
            trainable=False)

        self._extra_train_ops.append(moving_averages.assign_moving_average(
            moving_mean, mean, 0.9))
        self._extra_train_ops.append(moving_averages.assign_moving_average(
            moving_variance, variance, 0.9))
      elif self.mode == 'eval':
        mean = tf.get_variable(
            'moving_mean', params_shape, tf.float32,
            initializer=tf.constant_initializer(0.0, tf.float32),
            trainable=False)
        variance = tf.get_variable(
            'moving_variance', params_shape, tf.float32,
            initializer=tf.constant_initializer(1.0, tf.float32),
            trainable=False)
        tf.summary.histogram(mean.op.name, mean)
        tf.summary.histogram(variance.op.name, variance)
      else:
        mean, variance = tf.nn.moments(x[:-1,], [0, 1, 2], name='moments')

      # epsilon used to be 1e-5. Maybe 0.001 solves NaN problem in deeper net.
      y = tf.nn.batch_normalization(
          x, mean, variance, beta, gamma, 0.001)
      y.set_shape(x.get_shape())
      return y
```
I have several questions:

1. If `self.mode` is set as `mcbn`, the mean and variance will be fixed for a given `x` without randomness as:
```
mean, variance = tf.nn.moments(x[:-1,], [0, 1, 2], name='moments')
```
However, as shown in the paper Algorithm 1, they should be randomly estimated by batches in the training set.

2. During training, the mean and variance are:
```
 mean, variance = tf.nn.moments(x, [0, 1, 2], name='moments')
```
and they are used in:
```
y = tf.nn.batch_normalization(x, mean, variance, beta, gamma, 0.001)
```
It means the estimated and tracked `moving_mean` and `moving_variance` are nevered during training. I think this process is not same as vanilla training with batch normalization, while it is claimed that "The network is trained just as a regular BN network".

Looking forward to your reply.

Best regards,
lan-qing
